### ISSUE TITLE
Carbon doesn't take not-wellformed method precondition into account

### ISSUE THREAD

**Issue by tschoesi** [viperproject/carbon#388]
Carbon currently reports an error for [this assertion](https://github.com/viperproject/silver/blob/83ff132b847fd97ef20e1b622167de706d52a29a/src/test/resources/all/issues/silicon/0083.vpr#L21), and I think it shouldn't.

```
1. method test2(xs: Seq[Int], n: Int)
2.   requires 5 < n
3.   //:: ExpectedOutput(not.wellformed:seq.index.length)
4.   requires forall i: Int :: i in [0..n) ==> xs[i] != 0
5. {
6.   //:: ExpectedOutput(assert.failed:seq.index.length)
7.   //:: MissingOutput(assert.failed:seq.index.length, /silicon/issue/34/)
8.   assert xs[3] != 0
9. }
```

From the precondition on line 2 we know that `n>=6`. 
From the precondition on line 4 we know that for all `i < 6, xs[i] != 0` (and therefore that `3` is not out of bounds for `xs`).

So I think the assertion on line 8 should actually hold.

**Comment by gauravpartha** [comment]
I agree that Carbon should not report an error on line 8. Usually, Carbon takes the well-definedness of previous assertions into account, but for the pure quantifier ``forall i: Int :: body(i)`` the encoding is essentially:

```
if (*) {
  var k: Int //pick arbitrary value
   *** check that body(k) is well-defined ***
  assume false //kill the trace
}
assume forall i: Int :: body(i) //bring assumption into main trace
```

So, the potential well-definedness failure does not get propagated to the "main trace" (not just because the trace is killed; even if the trace were not killed, then there still is no single trace that contains all the information).

If you replace the assumption on line 4 by ```requires xs[3] != 0```, then the errors are reported as expected.

===Side remark===
On a more philosophical note, I would give a slightly different explanation for why the assertion on line 8 should hold. For any trace for which `inhale A` does not fail, it must hold that `A` is well-defined. In particular, well-definedness of `A` implies that every sequence lookup is well-defined (i.e., in the example, `i` is smaller than the length). In other words, I would argue directly via well-definedness instead of via satisfiability (but this is a very minor point).
